From 177d75e8c43be45a8383d794ce4084dd5d600a9e Mon Sep 17 00:00:00 2001
From: Evgeny Vereshchagin <evvers@ya.ru>
Date: Fri, 3 Nov 2023 03:13:36 +0000
Subject: [PATCH] core: compare RRs with zero-length rdata correctly

RRs where types are unknown to avahi can't be passed to
lexicographical_memcmp directly because it doesn't expect NULL pointers
and crashes. This patch prevents it from crashing by treating those NULL
pointers as empty strings and returning -1, 0 or 1. It in turn prevents
avahi from crashing on receiving probes with zero-length rdata if
unusual RRs have ever been added using avahi_entry_group_add_record.

This patch also fixes a bug where RRs with no remaining bytes were
deemed lexicographically later even though according to
https://datatracker.ietf.org/doc/html/rfc6762#section-8.2

The bytes of the raw uncompressed rdata are compared ...
until a byte is found whose value is greater than that of its counterpart
... or one of the resource records runs out of rdata (in which case,
the resource record which still has remaining data first is deemed
lexicographically later)

Closes https://github.com/lathiat/avahi/issues/493
---
 avahi-core/dns-test.c | 49 +++++++++++++++++++++++++++++++++++++++++++
 avahi-core/rr.c       | 22 +++++++++++++++----
 2 files changed, 67 insertions(+), 4 deletions(-)

diff --git a/avahi-core/dns-test.c b/avahi-core/dns-test.c
index ebe2305..c7116a2 100644
--- a/avahi-core/dns-test.c
+++ b/avahi-core/dns-test.c
@@ -33,6 +33,7 @@
 
 #include "dns.h"
 #include "log.h"
+#include "rr-util.h"
 #include "util.h"
 
 int main(AVAHI_GCC_UNUSED int argc, AVAHI_GCC_UNUSED char *argv[]) {
@@ -42,6 +43,7 @@ int main(AVAHI_GCC_UNUSED int argc, AVAHI_GCC_UNUSED char *argv[]) {
     AvahiRecord *r, *r2;
     uint8_t rdata[AVAHI_DNS_RDATA_MAX];
     size_t l;
+    int res;
 
     p = avahi_dns_packet_new(0);
 
@@ -109,5 +111,52 @@ int main(AVAHI_GCC_UNUSED int argc, AVAHI_GCC_UNUSED char *argv[]) {
     avahi_free(m);
     avahi_record_unref(r);
 
+    r = avahi_record_new_full("test", 77, 77, AVAHI_DEFAULT_TTL);
+    assert(r);
+
+    r2 = avahi_record_new_full("test", 77, 77, AVAHI_DEFAULT_TTL);
+    assert(r2);
+
+    // "" == ""
+    res = avahi_record_lexicographical_compare(r, r2);
+    assert(res == 0);
+
+    r->data.generic.data = avahi_memdup("WAT", r->data.generic.size = 3);
+    assert(r->data.generic.data);
+
+    // "WAT" > ""
+    res = avahi_record_lexicographical_compare(r, r2);
+    assert(res == 1);
+
+    // "" < "WAT"
+    res = avahi_record_lexicographical_compare(r2, r);
+    assert(res == -1);
+
+    r2->data.generic.data = avahi_memdup("WA", r2->data.generic.size = 2);
+    assert(r2->data.generic.data);
+
+    // "WAT" > "WA"
+    res = avahi_record_lexicographical_compare(r, r2);
+    assert(res == 1);
+
+    // "WA" < "WAT"
+    res = avahi_record_lexicographical_compare(r2, r);
+    assert(res == -1);
+
+    avahi_record_unref(r);
+
+    r = avahi_record_new_full("test", 77, 77, AVAHI_DEFAULT_TTL);
+    assert(r);
+
+    r->data.generic.data = avahi_memdup("WA", r->data.generic.size = 2);
+    assert(r->data.generic.data);
+
+    // "WA" == "WA"
+    res = avahi_record_lexicographical_compare(r, r2);
+    assert(res == 0);
+
+    avahi_record_unref(r);
+    avahi_record_unref(r2);
+
     return 0;
 }
diff --git a/avahi-core/rr.c b/avahi-core/rr.c
index 9c04ebb..7824166 100644
--- a/avahi-core/rr.c
+++ b/avahi-core/rr.c
@@ -547,7 +547,7 @@ static int lexicographical_memcmp(const void* a, size_t al, const void* b, size_
     if (al == bl)
         return 0;
     else
-        return al == c ? 1 : -1;
+        return al == c ? -1 : 1;
 }
 
 static int uint16_cmp(uint16_t a, uint16_t b) {
@@ -640,9 +640,23 @@ int avahi_record_lexicographical_compare(AvahiRecord *a, AvahiRecord *b) {
         case AVAHI_DNS_TYPE_AAAA:
             return memcmp(&a->data.aaaa.address, &b->data.aaaa.address, sizeof(AvahiIPv6Address));
 
-        default:
-            return lexicographical_memcmp(a->data.generic.data, a->data.generic.size,
-                                          b->data.generic.data, b->data.generic.size);
+        default: {
+            size_t asize, bsize;
+
+            asize = a->data.generic.size;
+            bsize = b->data.generic.size;
+
+            if (asize && bsize)
+                r = lexicographical_memcmp(a->data.generic.data, asize, b->data.generic.data, bsize);
+            else if (asize && !bsize)
+                r = 1;
+            else if (!asize && bsize)
+                r = -1;
+            else
+                r = 0;
+
+            return r;
+        }
     }
 
 
